stablehlo/transforms/StablehloLegalizeDeprecatedOpsPatterns.td

```diff
@@ -0,0 +1,56 @@
+/* Copyright 2024 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This is the canonicalize pattern definition file.
+
+include "mlir/IR/OpBase.td"
+include "stablehlo/dialect/StablehloOps.td"
+
+def UnaryToBinaryEinsumEq : NativeCodeCall<
+  "$_builder.getStringAttr(\",\" + $0.getValue().str())">;
+
+def GetI64DenseElementsAttr : NativeCodeCall<
+  "$0.mapValues($_builder.getI64Type(), [](llvm::APInt x) { return x.sext(64); })">;
+
+def GetDefaultDotDimensionNumbers : NativeCodeCall<
+  "getDefaultDotDimensionNumbers($0)">;
+
+def GetDefaultBroadcastDimensions : NativeCodeCall<
+  "getBroadcastDimensions(cast<RankedTensorType>($0.getType()), cast<DenseI64ArrayAttr>($1))">;
+
+
+// Here, the element type can be any integer or float type. But, note that only
+// 32 bit integers are supported for the value.
+class GetScalarOfType<int value> : NativeCodeCall<
+  "getScalarOfType(getElementTypeOrSelf($0)," # value # ")">;
+
+def BroadcastToBroadcastInDim : Pat<
+  (StableHLO_BroadcastOp:$result $operand, $broadcast_sizes),
+  (StableHLO_BroadcastInDimOp $operand, (GetDefaultBroadcastDimensions $result, $broadcast_sizes))>;
+
+def CreateTokenToAfterAll : Pat<
+  (StableHLO_CreateTokenOp),
+  (StableHLO_AfterAllOp (NativeCodeCall<"ValueRange{}">))>;
```

Comment:
Curious as to why the NativeCodeCall is needed?

Reply:
If the alternative is specifying nothing, then that doesn't work. DRR has a heuristic to pick which build method to call (based on if it implements type inference etc), and for this op is requires an operand range to be specified. Omitting the range is a custom builder we added:

```
/usr/local/google/home/gleasonk/Coding/openxla/stablehlo/stablehlo/transforms/StablehloLegalizeDeprecatedOpsPatterns.td:43:5: error: op 'stablehlo.after_all' argument number mismatch: 0 in pattern vs. 1 in definition
def CreateTokenToAfterAll : Pat<
    ^
ninja: build stopped: subcommand failed.
```

Reply:
Ah that makes sense. I guess this is an example of how variadic arguments do not always play well with DRR.